### PR TITLE
add support for adding relative URLs to the index (fixes #121)

### LINKED ISSUE #121: add support for replicated repos
The current implementation of helms3 does not support S3 bucket replicated content. Ideally we would be able to leverage helm's ability to fetch charts using relative URLs (https://github.com/helm/helm/issues/3057)

It may in fact make sense to make all URLs relative, but for now my proposal is to make toggling this feature using a flag.

### PR DESCRIPTION
**Pull request by bashims**
Signed-off-by: Benjamin Ash <bash@intelerad.com>

### CODE REVIEW

**Comment by bashims** [comment]
@hypnoglow any chance you could look at this PR?

Thanks!

**Comment by davidkarlsen** [comment]
I think this is an elegant solution which obsoletes https://github.com/hypnoglow/helm-s3/pull/56

**Comment by davidkarlsen** [comment]
@bashims this repo seems dead - how can one apply your PR and run with that version in helm?

**Comment by bashims** [comment]
@davidkarlsen yeah not entirely sure what is going on here. You can install from my forked release if you want - I tagged it as v0.9.3

Installation instructions:

`helm plugin install https://github.com/bashims/helm-s3.git --version 0.9.3`

File: hack/integration-tests-local.sh [hunk 1]
```diff
@@ -9,32 +9,45 @@ export AWS_DEFAULT_REGION=us-east-1
 export AWS_ENDPOINT=localhost:9000
 export AWS_DISABLE_SSL=true
 
-docker container run -d --name helm-s3-minio \
+DOCKER_NAME='helm-s3-minio'
+
+cleanup() {
+  if $(docker container ls | grep -q "${DOCKER_NAME}\$") ; then
```
Comment (hypnoglow) [inline hack/integration-tests-local.sh:15]: We should omit `$()` here:

```suggestion
  if docker container ls | grep -q "${DOCKER_NAME}$" ; then
```

File: hack/integration-tests-local.sh [hunk 2]
```diff
@@ -9,32 +9,45 @@ export AWS_DEFAULT_REGION=us-east-1
...
 if [ ! -x "$(which mc 2>/dev/null)" ]; then
-    go get -d github.com/minio/mc
-    (cd ${MCGOPATH} && make)
+    pushd /tmp > /dev/null
+    go get github.com/minio/mc
+    popd > /dev/null
 fi
 
-PATH=${MCGOPATH}:${PATH}
+sleep 3
```
Comment (hypnoglow) [inline hack/integration-tests-local.sh:42]: Could you add a comment why we need to sleep here?

File: hack/integration-tests-local.sh [hunk 3]
```diff
@@ -9,32 +9,45 @@ export AWS_DEFAULT_REGION=us-east-1
...
+cleanup() {
+  if $(docker container ls | grep -q "${DOCKER_NAME}\$") ; then
+    docker container rm --force --volumes "${DOCKER_NAME}" || :
+  fi
+}
+
+cleanup
+
+on_exit() {
+  if [ -z "${SKIP_CLEANUP}" ]; then
```
Comment (hypnoglow) [inline hack/integration-tests-local.sh:23]: This test now fails without this variable set: 

```
./hack/integration-tests-local.sh: line 26: SKIP_CLEANUP: unbound variable
```

```suggestion
  if [ -z "${SKIP_CLEANUP:-}" ]; then
```

File: hack/integration-tests-local.sh [hunk 4]
```diff
@@ -9,32 +9,45 @@ export AWS_DEFAULT_REGION=us-east-1
 export AWS_ENDPOINT=localhost:9000
 export AWS_DISABLE_SSL=true
 
-docker container run -d --name helm-s3-minio \
+DOCKER_NAME='helm-s3-minio'
+
+cleanup() {
+  if $(docker container ls | grep -q "${DOCKER_NAME}\$") ; then
+    docker container rm --force --volumes "${DOCKER_NAME}" || :
```
Comment (hypnoglow) [inline hack/integration-tests-local.sh:16]:
```suggestion
    docker container rm --force --volumes "${DOCKER_NAME}" &>/dev/null || :
```

File: hack/integration-tests-local.sh [hunk 5]
```diff
@@ -9,32 +9,45 @@ export AWS_DEFAULT_REGION=us-east-1
...
+cleanup
+
+on_exit() {
+  if [ -z "${SKIP_CLEANUP}" ]; then
+    cleanup
+  fi
+}
+trap on_exit EXIT
+
+docker container run -d --rm --name helm-s3-minio \
```
Comment (hypnoglow) [inline hack/integration-tests-local.sh:29]:
```suggestion
docker container run -d --rm --name "${DOCKER_NAME}" \
```

File: tests/e2e/push_test.go [hunk 6]
```diff
@@ -206,17 +182,70 @@ func TestPushForceAndIgnoreIfExists(t *testing.T) {
...
+		t.Fatalf("Unexpected error: %v", err)
+	}
+
+	expected := []string{filename}
+	if diff := cmp.Diff(expected, v.URLs); diff != "" {
+		t.Errorf("mismatch (-want +got):\n%s", diff)
+	}
+
+	os.Chdir(tmpdir)
+	cmd, stdout, stderr = command(fmt.Sprintf("helm fetch %s/%s --version %s", name, chartName, chartVer))
```
Comment (hypnoglow) [inline tests/e2e/push_test.go:242]: `os.Chdir` can cause problems.

Could you please either add `--destination` flag to helm command or use something like

```
	wd, _ := os.Getwd()
	os.Chdir(tmpdir)
	defer os.Chdir(wd)
```

**Review by hypnoglow — CHANGES_REQUESTED** [review]
Thanks a lot for this contribution!

Overall looks fine, a few nits on test changes.

**Comment by hypnoglow** [comment]
Tested it locally, works fine, thanks.

File: hack/integration-tests-local.sh [hunk 7]
```diff
@@ -1,5 +1,5 @@
 #!/usr/bin/env bash
-set -uo pipefail
+set -x -e -uo pipefail
```
[the remark this answers is not in the document]
Comment (bashims) [reply hack/integration-tests-local.sh:2]: fixed

File: hack/integration-tests-local.sh (hunk 1 above)
Comment (bashims) [reply hack/integration-tests-local.sh:15]: fixed

**Comment by hypnoglow** [comment]
@bashims I don't see any changes in the code. Have you pushed them?

**Comment by bashims** [comment]
@hypnoglow just pushed

**Comment by hypnoglow** [comment]
Thank you!

**Pull request merged**